Source/CLI/rawcooked.1
.TH "RAWcooked" "1" "https://mediaarea.net/RAWcooked" "18.09beta1" "Bit\-by\-bit fidelity"
.\" Turn off justification for nroff.
.if n .ad l
.\" Turn off hyphenation.
.nh
.SH NAME
\fBRAWcooked\fR \- encode and decode audio\-visual RAW data with Matroska, FFV1 and FLAC
.SH SYNOPSIS
\fBrawcooked \fR[\fIoption\fR ...] (\fIfolder\fR | \fIfile\fR ...) [\fIoption\fR ...]
.SH DESCRIPTION
.TP
\fBRAWcooked\fR encodes audio\-visual RAW data into the Matroska container (MKV), using the video codec FFV1 for the image and the audio codec FLAC for the sound. The metadata accompanying the RAW data are preserved, and sidecar files, like MD5, LUT or XML, can be added into the Matroska container as attachments. This allows for the management of these audio\-visual file formats in an effective and transparent way (e.g. native playback in VLC), while saving typically between one to two thirds of the needed storage, and speeding up file writing and reading (e.g. to/from harddisk, over network or for backup on LTO).
.TP
When needed, the uncompressed source is retrieved \fBbit\-by\-bit\fR, in a manner faster than uncompressed sources. 
.TP
.I folder
Encodes with the FFV1 video codec all single\-image video files in the \fIfolder\fR, encodes with the FLAC audio codec all audio files in the \fIfolder\fR, and muxes these into a Matroska container (.mkv).
.br
The filenames of the single image files must end with a numbered sequence. \fBRAWcooked\fR will generate the regex to parse in the correct order all the frames in the folder.
.TP
.I file
.B contains RAW data (e.g. a .dpx or .wav file):
.br
Encodes with the FFV1 video codec all single\-image video files in the folder containing the \fIfile\fR, encodes with the FLAC audio codec all audio files in the folder containing the \fIfile\fR, and muxes these into a Matroska container (.mkv).
.br
The filenames usually end with a numbered sequence. Enter one frame and the tool will generate the regex to parse all the frames in the folder.
.TP
.I file
.B is a Matroska container (.mkv):
.br
Decodes back the video and the audio of \fIfile\fR to its original formats. All metadata accompanying the original data are preserved bit\-by\-bit.
.SH OPTIONS
.SS GENERAL OPTIONS
.TP
.B \-\-help \fR|\fB \-h
Displays a help message.
.TP
.B \-\-version
Displays the installed version.
.TP
.B \-\-attachment\-max\-size \fIvalue\fR | \fB\-s \fIvalue
Set maximum size of attachment to \fIvalue\fR (in bytes).
.br
Default value is \fI1048576\fR.
.TP
.B \-\-check \fIvalue\fR
Do or don't do costly (in terms of analysis duration or bytes read) checks.
.br
\fIpartial\fR (or \fI0\fR) is quicker but may lead to partial reversibility with non conform files.
.br
\fIfull\fR (or \fI1\fR) is slower but  guaranties reversability with e.g. DPX files with non zero padding bits.
.br
Default value is 'partial'.
.TP
.B \-\-display\-command \fR|\fB \-d
When an external encoder/decoder is used, display the command to launch instead of launching it.
.TP
.B \-\-output\-name \fIvalue\fR | \fB\-o \fIvalue
Set the name of the output file or folder to \fIvalue\fR.
.br
Default value is \fI${Input}.mkv\fR (if input is a folder) or \fI${Input}.RAWcooked\fR (if input is a file).
.TP
.B \-\-rawcooked\-file\-name \fIvalue\fR | \fB\-r \fIvalue
Set (encoding) or get (decoding) the name of the \fBRAWcooked\fR reversibility data file to \fIvalue\fR.
.br
Default name is \fI${Input}.rawcooked_reversibility_data\fR
.br
\fBNote:\fR If the \fBRAWcooked\fR reversibility data file is included in the output A/V file during the encoding, this file is deleted after encoding.
.br
\fBNote:\fR Not yet implemented for decoding.
.TP
.B \-\-quiet
Do not show information related to RAWcooked.
.br
External encoder/decoder may need an additional option.
.SS INPUT RELATED OPTIONS
.TP
.B \-\-file
Unlock compression of files (e.g. a .dpx or .wav).
.TP
.B \-framerate \fIvalue
Force video frame rate to \fIvalue\fR.
.br
Default value is the one found in the image files if available, otherwise \fI24\fR.
.SS ENCODING RELATED OPTIONS
.TP
.B \-c:a \fIvalue
Force the audio encoding format to \fIvalue\fR: \fIcopy\fR (copy PCM to PCM, without modification), \fIflac\fR.
.br
Default value is \fIflac\fR.
.TP
.B \-c:v \fIvalue
Force the video encoding format \fIvalue\fR: only \fIffv1\fR is currently allowed.
.br
Default value is \fIffv1\fR.
.TP
.B \-coder \fIvalue
If video encoding format is \fIffv1\fR, set the coder to \fIvalue\fR: \fI0\fR (Golomb-Rice), \fI1\fR (Range Coder), \fI2\fR (Range Coder with custom state transition table).
.br
Default value is \fI1\fR.
.TP
.B \-context \fIvalue
If video encoding format is \fIffv1\fR, set the context to \fIvalue\fR: \fI0\fR (small), \fI1\fR (large).
.br
Default value is \fI0\fR.
.TP
.B \-format \fIvalue
Set the container format to \fIvalue\fR: only \fImatroska\fR is currently allowed.
.br
Default value is \fImatroska\fR.
.TP
.B \-g \fIvalue
If video encoding format is \fIffv1\fR, set the GOP size to \fIvalue\fR\: any integer >=1.
.br
Default value is \fI1\fR.
.TP
.B \-level \fIvalue
If video encoding format is \fIffv1\fR, set the version to \fIvalue\fR: \fI0\fR, \fI1\fR, \fI3\fR.
.br
Default value is \fI3\fR.
.TP
.B \-slicecrc \fIvalue
If video encoding format is \fIffv1\fR, set the CRC to \fIvalue\fR: \fI0\fR (CRC not present), \fI1\fR (CRC present).
.br
Default value is \fI1\fR.
.TP
.B \-slices \fIvalue
If video encoding format is \fIffv1\fR, set the count of slices to \fIvalue\fR: any integer >=1 and making sense (2, 4, 6, 9, 16, 24...).
.br
Default value is between \fI16\fR and \fI512\fR, depending on video frame size and depth.
.SH COPYRIGHT
Copyright (c) 2018 MediaArea.net SARL & AV Preservation by reto.ch
.SH LICENSE
\fBRAWcooked\fR is released under a BSD License.
.SH DISCLAIMER
\fBRAWcooked\fR is provided "as is" without warranty or support of any kind.


debian/changelog
rawcooked (18.09beta1-1) unstable; urgency=low

  * See History.txt for more info and real dates

 -- Jerome Martinez <info@mediaarea.net>  Sun, 01 Jan 2017 01:01:01 +0000



Source/Lib/RAWcooked/RAWcooked.cpp
/*  Copyright (c) MediaArea.net SARL & AV Preservation by reto.ch.
 *
 *  Use of this source code is governed by a BSD-style license that can
 *  be found in the License.html file in the root of the source tree.
 */

//---------------------------------------------------------------------------
#include "Lib/RAWcooked/RAWcooked.h"
#include "Lib/Config.h"
#include "zlib.h"
#include <cstring>
using namespace std;
//---------------------------------------------------------------------------

// Library name and version
const char* LibraryName = "RAWcooked";
const char* LibraryVersion = "18.09beta1";

// EBML
static const uint32_t Name_EBML = 0x0A45DFA3;
static const uint32_t Name_EBML_Doctype = 0x0282;
static const uint32_t Name_EBML_DoctypeVersion = 0x0287;
static const uint32_t Name_EBML_DoctypeReadVersion = 0x0285;

// Top level
static const uint32_t Name_RawCookedSegment = 0x7273; // "rs", RAWcooked Segment part
static const uint32_t Name_RawCookedTrack = 0x7274;   // "rt", RAWcooked Track part
static const uint32_t Name_RawCookedBlock = 0x7262;   // "rb", RAWcooked BlockGroup part
// File data
static const uint32_t Name_RawCooked_BeforeData = 0x01;
static const uint32_t Name_RawCooked_AfterData = 0x02;
static const uint32_t Name_RawCooked_MaskBaseBeforeData = 0x03;     // In BlockGroup only
static const uint32_t Name_RawCooked_MaskBaseAfterData = 0x04;      // In BlockGroup only
static const uint32_t Name_RawCooked_MaskAdditionBeforeData = 0x03; // In Track only
static const uint32_t Name_RawCooked_MaskAdditionAfterData = 0x04;  // In Track only
static const uint32_t Name_RawCooked_InData = 0x05;
static const uint32_t Name_RawCooked_MaskBaseInData = 0x06;         // In BlockGroup only
static const uint32_t Name_RawCooked_MaskAdditionInData = 0x06;     // In Track only
// File text metadata
static const uint32_t Name_RawCooked_FileName = 0x10;
static const uint32_t Name_RawCooked_MaskBaseFileName = 0x11;     // In BlockGroup only
static const uint32_t Name_RawCooked_MaskAdditionFileName = 0x11; // In Track only
// File stats
static const uint32_t Name_RawCooked_FileMD5 = 0x20;
static const uint32_t Name_RawCooked_FileSHA1 = 0x21;
static const uint32_t Name_RawCooked_FileSHA256 = 0x22;
// File number metadata
static const uint32_t Name_RawCooked_FileSize = 0x30;
static const uint32_t Name_RawCooked_MaskBaseFileSize = 0x31;     // In BlockGroup only
static const uint32_t Name_RawCooked_MaskAdditionFileSize = 0x32; // In Track only
// Global information
static const uint32_t Name_RawCooked_LibraryName = 0x70;
static const uint32_t Name_RawCooked_LibraryVersion = 0x71;
static const uint32_t Name_RawCooked_PathSeparator = 0x72;
// Parameters
static const char* DocType = "rawcooked";
static const uint8_t DocTypeVersion = 1;
static const uint8_t DocTypeReadVersion = 1;

//---------------------------------------------------------------------------
static size_t Size_EB(uint32_t Name, uint64_t Size)
{
    size_t N_l = 1;
    while (Name >> (N_l * 7))
        N_l++;

    size_t S_l = 1;
    while (Size >> (S_l * 7))
        S_l++;

    return N_l + S_l + Size;
}

//---------------------------------------------------------------------------
static size_t Size_EB(uint64_t Size)
{
    size_t S_l = 1;
    while (Size >> (S_l * 7))
        S_l++;
    return S_l;
}

//---------------------------------------------------------------------------
static size_t Size_Number(uint64_t Number)
{
    size_t S_l = 1;
    while (Number >> (S_l * 8))
        S_l++;
    return S_l;
}

//---------------------------------------------------------------------------
static void Put_EB(unsigned char* Buffer, size_t& Offset, uint32_t Name, uint64_t Size)
{
    size_t N_l = 1;
    while (Name >> (N_l * 7))
        N_l++;
    uint32_t N2 = Name | (((uint32_t)1) << (N_l * 7));
    while (N_l)
    {
        Buffer[Offset] = (uint8_t)(N2 >> ((N_l - 1)*8));
        Offset++;
        N_l--;
    }

    size_t S_l = 1;
    while (Size >> (S_l * 7))
        S_l++;
    uint64_t S2 = Size | (((uint64_t)1) << (S_l * 7));
    while (S_l)
    {
        Buffer[Offset] = (uint8_t)(S2 >> ((S_l - 1)*8));
        Offset++;
        S_l--;
    }
}

//---------------------------------------------------------------------------
static void Put_EB(unsigned char* Buffer, size_t& Offset, uint64_t Size)
{
    size_t S_l = 1;
    while (Size >> (S_l * 7))
        S_l++;
    uint64_t S2 = Size | (((uint64_t)1) << (S_l * 7));
    while (S_l)
    {
        Buffer[Offset] = (uint8_t)(S2 >> ((S_l - 1) * 8));
        Offset++;
        S_l--;
    }
}

//---------------------------------------------------------------------------
static void Put_Number(unsigned char* Buffer, size_t& Offset, uint64_t Number)
{
    size_t S_l = 1;
    while (Number >> (S_l * 8))
        S_l++;
    while (S_l)
    {
        Buffer[Offset] = (uint8_t)(Number >> ((S_l - 1) * 8));
        Offset++;
        S_l--;
    }
}

//---------------------------------------------------------------------------
rawcooked::rawcooked() :
    Unique(false),
    F(NULL),
    In(NULL),
    In_Size(0),
    FileSize((uint64_t)-1),
    BlockCount(0)
{
}

//---------------------------------------------------------------------------
rawcooked::~rawcooked()
{
    Close();
}

//---------------------------------------------------------------------------
void rawcooked::Parse()
{
    // Cross-platform support
    // RAWcooked file format supports setting of the path separator but
    // we currently set all to "/", which is supported by both Windows and Unix based platforms
    // On Windows, we convert "\" to "/" as both are considered as path separators and Unix-based  systems don't consider "\" as a path separator
    // If not doing this, files are not considered as in a sub-directory when encoded with a Windows platform then decoded with a Unix-based platform.
    #if defined(_WIN32) || defined(_WINDOWS)
        string::size_type i = 0;
        while ((i = OutputFileName.find('\\', i)) != string::npos)
            OutputFileName[i++] = '/';
    #endif

    // Create or Use mask
    uint8_t* FileName = (uint8_t*)OutputFileName.c_str();
    size_t FileName_Size = OutputFileName.size();
    uint8_t* ToStore_FileName = FileName;
    uint8_t* ToStore_Before = Before;
    uint8_t* ToStore_After = After;
    uint8_t* ToStore_In = In;
    bool IsUsingMask_FileName = false;
    bool IsUsingMask_Before = false;
    bool IsUsingMask_After = false;
    if (!Unique)
    {
        if (!BlockCount)
        {
            if (FileName && FileName_Size)
            {
                FirstFrame_FileName = new uint8_t[FileName_Size];
                memcpy(FirstFrame_FileName, FileName, FileName_Size);
                FirstFrame_FileName_Size = FileName_Size;
            }
            else
            {
                FirstFrame_FileName = NULL;
                FirstFrame_FileName_Size = 0;
            }
            if (Before && Before_Size)
            {
                FirstFrame_Before = new uint8_t[Before_Size];
                memcpy(FirstFrame_Before, Before, Before_Size);
                FirstFrame_Before_Size = Before_Size;
            }
            else
            {
                FirstFrame_Before = NULL;
                FirstFrame_Before_Size = 0;
            }
            if (After && After_Size)
            {
                FirstFrame_After = new uint8_t[After_Size];
                memcpy(FirstFrame_After, After, After_Size);
                FirstFrame_After_Size = After_Size;
            }
            else
            {
                FirstFrame_After = NULL;
                FirstFrame_After_Size = 0;
            }
        }
        if (FileName && FirstFrame_FileName)
        {
            ToStore_FileName = new uint8_t[FileName_Size];
            memcpy(ToStore_FileName, FileName, FileName_Size);
            for (size_t i = 0; i < FileName_Size || i < FirstFrame_FileName_Size; i++)
                ToStore_FileName[i] -= FirstFrame_FileName[i];
            IsUsingMask_FileName = true;
        }
        if (Before && FirstFrame_Before)
        {
            ToStore_Before = new uint8_t[Before_Size];
            memcpy(ToStore_Before, Before, Before_Size);
            for (size_t i = 0; i < Before_Size || i < FirstFrame_Before_Size; i++)
                ToStore_Before[i] -= FirstFrame_Before[i];
            IsUsingMask_Before = true;
        }
        if (After && FirstFrame_After)
        {
            ToStore_After = new uint8_t[After_Size];
            memcpy(ToStore_After, After, After_Size);
            for (size_t i = 0; i < After_Size || i < FirstFrame_After_Size; i++)
                ToStore_After[i] -= FirstFrame_After[i];
            IsUsingMask_After = true;
        }
    }

    // Test if it can be compressed
    uint8_t* Compressed_MaskFileName = NULL;
    uLongf Compressed_MaskFileName_Size = 0;
    if (!Unique && FirstFrame_FileName)
    {
        Compressed_MaskFileName = new uint8_t[FirstFrame_FileName_Size];
        Compressed_MaskFileName_Size = (uLongf)FirstFrame_FileName_Size;
        if (compress((Bytef*)Compressed_MaskFileName, &Compressed_MaskFileName_Size, (const Bytef*)FirstFrame_FileName, (uLong)FirstFrame_FileName_Size) < 0)
        {
            Compressed_MaskFileName = FirstFrame_FileName;
            Compressed_MaskFileName_Size = 0;
        }
    }
    uint8_t* Compressed_FileName = new uint8_t[FileName_Size];
    uLongf Compressed_FileName_Size = (uLongf)FileName_Size;
    if (compress((Bytef*)Compressed_FileName, &Compressed_FileName_Size, (const Bytef*)ToStore_FileName, (uLong)FileName_Size) < 0)
    {
        Compressed_FileName = ToStore_FileName;
        Compressed_FileName_Size = 0;
    }
    uint8_t* Compressed_MaskBefore = NULL;
    uLongf Compressed_MaskBefore_Size = 0;
    if (!Unique && FirstFrame_Before)
    {
        Compressed_MaskBefore = new uint8_t[FirstFrame_Before_Size];
        Compressed_MaskBefore_Size = (uLongf)FirstFrame_Before_Size;
        if (compress((Bytef*)Compressed_MaskBefore, &Compressed_MaskBefore_Size, (const Bytef*)FirstFrame_Before, (uLong)FirstFrame_Before_Size) < 0)
        {
            Compressed_MaskBefore = FirstFrame_Before;
            Compressed_MaskBefore_Size = 0;
        }
    }
    uint8_t* Compressed_Before = new uint8_t[Before_Size];
    uLongf Compressed_Before_Size = (uLongf)Before_Size;
    if (compress((Bytef*)Compressed_Before, &Compressed_Before_Size, (const Bytef*)ToStore_Before, (uLong)Before_Size) < 0)
    {
        Compressed_Before = ToStore_Before;
        Compressed_Before_Size = 0;
    }
    uint8_t* Compressed_MaskAfter = NULL;
    uLongf Compressed_MaskAfter_Size = 0;
    if (!Unique && FirstFrame_After)
    {
        Compressed_MaskAfter = new uint8_t[FirstFrame_After_Size];
        Compressed_MaskAfter_Size = (uLongf)FirstFrame_After_Size;
        if (compress((Bytef*)Compressed_MaskAfter, &Compressed_MaskAfter_Size, (const Bytef*)FirstFrame_After, (uLong)FirstFrame_After_Size) < 0)
        {
            Compressed_MaskAfter = FirstFrame_After;
            Compressed_MaskAfter_Size = 0;
        }
    }
    uint8_t* Compressed_After = new uint8_t[After_Size];
    uLongf Compressed_After_Size = (uLongf)After_Size;
    if (compress((Bytef*)Compressed_After, &Compressed_After_Size, (const Bytef*)ToStore_After, (uLong)After_Size) < 0)
    {
        Compressed_After = ToStore_After;
        Compressed_After_Size = 0;
    }
    uint8_t* Compressed_In = new uint8_t[In_Size];
    uLongf Compressed_In_Size = (uLongf)In_Size;
    if (compress((Bytef*)Compressed_In, &Compressed_In_Size, (const Bytef*)ToStore_In, (uLong)In_Size) < 0)
    {
        Compressed_In = ToStore_In;
        Compressed_In_Size = 0;
    }

    uint64_t Out_Size = 0;

    // Size computing - EBML header
    uint64_t EBML_Size = 0;
    uint64_t EBML_DocType_Size;
    if (!F)
    {
        EBML_DocType_Size = strlen(DocType);
        EBML_Size += Size_EB(Name_EBML_Doctype, EBML_DocType_Size);
        EBML_Size += Size_EB(Name_EBML_DoctypeVersion, 1);
        EBML_Size += Size_EB(Name_EBML_DoctypeReadVersion, 1);
    }

    // Size computing - Library name/version
    uint64_t Segment_Size = 0;
    uint64_t LibraryName_Size;
    uint64_t LibraryVersion_Size;
    if (!F)
    {
        LibraryName_Size = strlen(LibraryName);
        Segment_Size += Size_EB(Name_RawCooked_LibraryName, LibraryName_Size);
        LibraryVersion_Size = strlen(LibraryVersion);
        Segment_Size += Size_EB(Name_RawCooked_LibraryVersion, LibraryVersion_Size);
    }

    // Size computing - Track part
    uint64_t Track_Size = 0;
    if (!BlockCount)
    {
        if (!Unique && FirstFrame_FileName)
            Track_Size += Size_EB(Name_RawCooked_MaskBaseFileName, Size_EB(Compressed_MaskFileName_Size ? FirstFrame_FileName_Size : 0) + (Compressed_MaskFileName_Size ? Compressed_MaskFileName_Size : FirstFrame_FileName_Size));
        if (!Unique && FirstFrame_Before)
            Track_Size += Size_EB(Name_RawCooked_MaskBaseBeforeData, Size_EB(Compressed_MaskBefore_Size ? FirstFrame_Before_Size : 0) + (Compressed_MaskBefore_Size ? Compressed_MaskBefore_Size : FirstFrame_Before_Size));
        if (!Unique && FirstFrame_After)
            Track_Size += Size_EB(Name_RawCooked_MaskBaseAfterData, Size_EB(Compressed_MaskAfter_Size ? FirstFrame_After_Size : 0) + (Compressed_MaskAfter_Size ? Compressed_MaskAfter_Size : FirstFrame_After_Size));
    }

    // Size computing - Block part
    uint64_t Block_Size = 0;
    uint64_t BeforeData_Size = 0;
    uint64_t FileNameData_Size = 0;
    if (Compressed_FileName_Size || FileName_Size)
    {
        FileNameData_Size = Size_EB(Compressed_FileName_Size ? FileName_Size : 0) + (Compressed_FileName_Size ? Compressed_FileName_Size : FileName_Size); // size then data
        Block_Size += Size_EB(Name_RawCooked_FileName, FileNameData_Size);
    }
    if (Compressed_Before_Size || Before_Size)
    {
        BeforeData_Size = Size_EB(Compressed_Before_Size ? Before_Size : 0) + (Compressed_Before_Size ? Compressed_Before_Size : Before_Size); // size then data
        Block_Size += Size_EB(Name_RawCooked_BeforeData, BeforeData_Size);
    }
    uint64_t AfterData_Size = 0;
    if (Compressed_After_Size || After_Size)
    {
        AfterData_Size = Size_EB(Compressed_After_Size ? After_Size : 0) + (Compressed_After_Size ? Compressed_After_Size : After_Size); // size then data
        Block_Size += Size_EB(Name_RawCooked_AfterData, AfterData_Size);
    }
    uint64_t InData_Size = 0;
    if (Compressed_In_Size || In_Size)
    {
        InData_Size = Size_EB(Compressed_In_Size ? In_Size : 0) + (Compressed_In_Size ? Compressed_In_Size : In_Size); // size then data
        Block_Size += Size_EB(Name_RawCooked_InData, InData_Size);
    }
    if (FileSize != (uint64_t)-1)
    {
        Block_Size += Size_EB(Name_RawCooked_InData, Size_Number(FileSize));
    }

    // Case if the file is unique
    if (Unique)
    {
        Track_Size += Block_Size;
        Block_Size = 0;
    }

    // Size computing - Headers
    if (EBML_Size)
        Out_Size += Size_EB(Name_EBML, EBML_Size);
    if (Segment_Size)
        Out_Size += Size_EB(Name_RawCookedSegment, Segment_Size);
    if (Track_Size)
        Out_Size += Size_EB(Name_RawCookedTrack, Track_Size);
    if (Block_Size)
        Out_Size += Size_EB(Name_RawCookedBlock, Block_Size);

    // Fill
    uint8_t* Out = new uint8_t[Out_Size];
    size_t Out_Offset = 0;

    // Fill - EBML part
    if (EBML_Size)
    {
        Put_EB(Out, Out_Offset, Name_EBML, EBML_Size);
        Put_EB(Out, Out_Offset, Name_EBML_Doctype, EBML_DocType_Size);
        memcpy(Out + Out_Offset, DocType, EBML_DocType_Size);
        Out_Offset += EBML_DocType_Size;
        Put_EB(Out, Out_Offset, Name_EBML_DoctypeVersion, 1);
        Out[Out_Offset]=DocTypeVersion;
        Out_Offset += 1;
        Put_EB(Out, Out_Offset, Name_EBML_DoctypeReadVersion, 1);
        Out[Out_Offset] = DocTypeReadVersion;
        Out_Offset += 1;
    }

    // Fill - Segment part
    if (Segment_Size)
    {
        Put_EB(Out, Out_Offset, Name_RawCookedSegment, Segment_Size);
        Put_EB(Out, Out_Offset, Name_RawCooked_LibraryName, LibraryName_Size);
        memcpy(Out + Out_Offset, LibraryName, LibraryName_Size);
        Out_Offset += LibraryName_Size;
        Put_EB(Out, Out_Offset, Name_RawCooked_LibraryVersion, LibraryVersion_Size);
        memcpy(Out + Out_Offset, LibraryVersion, LibraryVersion_Size);
        Out_Offset += LibraryVersion_Size;
    }

    // Fill - Track part
    if (Track_Size)
    {
        Put_EB(Out, Out_Offset, Name_RawCookedTrack, Track_Size);
        if (!Unique && FirstFrame_FileName)
        {
            Put_EB(Out, Out_Offset, Name_RawCooked_MaskBaseFileName, Size_EB(Compressed_MaskFileName_Size ? FirstFrame_FileName_Size : 0) + (Compressed_MaskFileName_Size ? Compressed_MaskFileName_Size : FirstFrame_FileName_Size));
            Put_EB(Out, Out_Offset, Compressed_MaskFileName_Size ? FirstFrame_FileName_Size : 0);
            memcpy(Out + Out_Offset, Compressed_MaskFileName, (Compressed_MaskFileName_Size ? Compressed_MaskFileName_Size : FirstFrame_FileName_Size));
            Out_Offset += (Compressed_MaskFileName_Size ? Compressed_MaskFileName_Size : FirstFrame_FileName_Size);
        }
        if (!Unique && FirstFrame_Before)
        {
            Put_EB(Out, Out_Offset, Name_RawCooked_MaskBaseBeforeData, Size_EB(Compressed_MaskBefore_Size ? FirstFrame_Before_Size : 0) + (Compressed_MaskBefore_Size ? Compressed_MaskBefore_Size : FirstFrame_Before_Size));
            Put_EB(Out, Out_Offset, Compressed_MaskBefore_Size ? FirstFrame_Before_Size : 0);
            memcpy(Out + Out_Offset, Compressed_MaskBefore, (Compressed_MaskBefore_Size ? Compressed_MaskBefore_Size : FirstFrame_Before_Size));
            Out_Offset += (Compressed_MaskBefore_Size ? Compressed_MaskBefore_Size : FirstFrame_Before_Size);
        }
        if (!Unique && FirstFrame_After)
        {
            Put_EB(Out, Out_Offset, Name_RawCooked_MaskBaseAfterData, Size_EB(Compressed_MaskAfter_Size ? FirstFrame_After_Size : 0) + (Compressed_MaskAfter_Size ? Compressed_MaskAfter_Size : FirstFrame_After_Size));
            Put_EB(Out, Out_Offset, Compressed_MaskAfter_Size ? FirstFrame_After_Size : 0);
            memcpy(Out + Out_Offset, Compressed_MaskAfter, (Compressed_MaskAfter_Size ? Compressed_MaskAfter_Size : FirstFrame_After_Size));
            Out_Offset += (Compressed_MaskAfter_Size ? Compressed_MaskAfter_Size : FirstFrame_After_Size);
        }
    }

    // Fill - Block part
    if (Block_Size)
        Put_EB(Out, Out_Offset, Name_RawCookedBlock, Block_Size);
    if (FileNameData_Size)
    {
        Put_EB(Out, Out_Offset, IsUsingMask_FileName ? Name_RawCooked_MaskAdditionFileName : Name_RawCooked_FileName, FileNameData_Size);
        Put_EB(Out, Out_Offset, Compressed_FileName_Size ? FileName_Size : 0);
        memcpy(Out + Out_Offset, Compressed_FileName, (Compressed_FileName_Size ? Compressed_FileName_Size : FileName_Size));
        Out_Offset += (Compressed_FileName_Size ? Compressed_FileName_Size : FileName_Size);
    }
    if (BeforeData_Size)
    {
        Put_EB(Out, Out_Offset, IsUsingMask_Before? Name_RawCooked_MaskAdditionBeforeData : Name_RawCooked_BeforeData, BeforeData_Size);
        Put_EB(Out, Out_Offset, Compressed_Before_Size ? Before_Size : 0);
        memcpy(Out + Out_Offset, Compressed_Before, (Compressed_Before_Size ? Compressed_Before_Size : Before_Size));
        Out_Offset += (Compressed_Before_Size ? Compressed_Before_Size : Before_Size);
    }
    if (AfterData_Size)
    {
        Put_EB(Out, Out_Offset, IsUsingMask_After ? Name_RawCooked_MaskAdditionAfterData : Name_RawCooked_AfterData, AfterData_Size);
        Put_EB(Out, Out_Offset, Compressed_After_Size ? After_Size : 0);
        memcpy(Out + Out_Offset, Compressed_After, (Compressed_After_Size ? Compressed_After_Size : After_Size));
        Out_Offset += (Compressed_After_Size ? Compressed_After_Size : After_Size);
    }
    if (InData_Size)
    {
        Put_EB(Out, Out_Offset, Name_RawCooked_InData, InData_Size);
        Put_EB(Out, Out_Offset, Compressed_In_Size ? In_Size : 0);
        memcpy(Out + Out_Offset, Compressed_In, (Compressed_In_Size ? Compressed_In_Size : In_Size));
        Out_Offset += (Compressed_In_Size ? Compressed_In_Size : In_Size);
    }
    if (FileSize != (uint64_t)-1)
    {
        Put_EB(Out, Out_Offset, Name_RawCooked_FileSize, Size_Number(FileSize));
        Put_Number(Out, Out_Offset, FileSize);
    }

    // Write
    WriteToDisk(Out, Out_Size);
    BlockCount++;

    // Clean up
    if (Compressed_FileName_Size)
        delete[] Compressed_FileName;
    if (Compressed_Before_Size)
        delete[] Compressed_Before;
    if (Compressed_After_Size)
        delete[] Compressed_After;
    if (Compressed_In_Size)
        delete[] Compressed_In;
    if (IsUsingMask_FileName)
        delete[] ToStore_FileName;
    if (IsUsingMask_Before)
        delete[] ToStore_Before;
    if (IsUsingMask_After)
        delete[] ToStore_After;
}

//---------------------------------------------------------------------------
void rawcooked::ResetTrack()
{
    BlockCount = 0;
}

//---------------------------------------------------------------------------
void rawcooked::Close()
{
    if (F)
    {
        fclose(F);
        F = NULL;
    }
}

//---------------------------------------------------------------------------
void rawcooked::WriteToDisk(uint8_t* Buffer, size_t Buffer_Size)
{
    if (!F)
    {
        #ifdef _MSC_VER
            #pragma warning(disable:4996)// _CRT_SECURE_NO_WARNINGS
        #endif
        F = fopen(FileName.c_str(), "wb");
        #ifdef _MSC_VER
            #pragma warning(default:4996)// _CRT_SECURE_NO_WARNINGS
        #endif
    }
    fwrite(Buffer, Buffer_Size, 1, F);
}
